File: src/truthound_dashboard/schemas/scan.py
"""PII scan-related Pydantic schemas.

This module defines schemas for PII scan API operations using th.scan().

The scan functionality detects personally identifiable information (PII)
in datasets and checks compliance with privacy regulations (GDPR, CCPA, LGPD).
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import Field

from .base import BaseSchema, IDMixin, ListResponseWrapper


class Regulation(str, Enum):
    """Supported privacy regulations for compliance checking."""

    GDPR = "gdpr"
    CCPA = "ccpa"
    LGPD = "lgpd"


# Type alias for regulation literal
RegulationLiteral = Literal["gdpr", "ccpa", "lgpd"]

# PII type categories commonly detected
PII_TYPES = [
    "email",
    "phone",
    "ssn",
    "credit_card",
    "ip_address",
    "date_of_birth",
    "address",
    "name",
    "passport",
    "driver_license",
    "national_id",
    "bank_account",
    "medical_record",
    "biometric",
]


class PIIScanRequest(BaseSchema):
    """Request to run PII scan on a data source.

    This schema maps to truthound's th.scan() parameters for maximum flexibility.
    All optional parameters default to None to use truthound's defaults.
    """

    # Column filtering
    columns: list[str] | None = Field(
        default=None,
        description="Columns to scan. If None, all columns are scanned.",
        examples=[["email", "phone", "ssn"]],
    )

    # Regulation compliance checking
    regulations: list[RegulationLiteral] | None = Field(
        default=None,
        description="Privacy regulations to check compliance: gdpr, ccpa, lgpd",
        examples=[["gdpr", "ccpa"]],
    )

    # Confidence threshold
    min_confidence: float = Field(
        default=0.8,
        ge=0.0,
        le=1.0,
        description="Minimum confidence threshold for PII detection (0.0-1.0)",
        examples=[0.8, 0.9],
    )


class PIIFinding(BaseSchema):
    """Single PII finding detected in a column.

    Represents one type of PII detected within a specific column,
    including confidence score and sample information.
    """

    column: str = Field(..., description="Column where PII was detected")
    pii_type: str = Field(
        ...,
        description="Type of PII detected",
        examples=["email", "ssn", "phone", "credit_card"],
    )
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence score for this detection (0.0-1.0)",
    )
    sample_count: int = Field(
        ...,
        ge=0,
        description="Number of values matching this PII type",
    )
    sample_values: list[str] | None = Field(
        default=None,
        description="Sample values that matched (redacted for privacy)",
    )


class RegulationViolation(BaseSchema):
    """Regulation compliance violation.

    Represents a violation of a specific privacy regulation
    detected in the scanned data.
    """

    regulation: RegulationLiteral = Field(
        ...,
        description="Violated regulation",
    )
    column: str = Field(
        ...,
        description="Column with violation",
    )
    pii_type: str = Field(
        ...,
        description="Type of PII causing the violation",
    )
    message: str = Field(
        ...,
        description="Human-readable violation description",
    )
    severity: Literal["low", "medium", "high", "critical"] = Field(
        default="high",
        description="Severity level of the violation",
    )


class PIIScanSummary(BaseSchema):
    """Summary statistics for a PII scan run."""

    total_columns_scanned: int = Field(
        default=0,
        ge=0,
        description="Total number of columns scanned",
    )
    columns_with_pii: int = Field(
        default=0,
        ge=0,
        description="Number of columns containing PII",
    )
    total_findings: int = Field(
        default=0,
        ge=0,
        description="Total number of PII findings",
    )
    has_violations: bool = Field(
        default=False,
        description="Whether any regulation violations were found",
    )
    total_violations: int = Field(
        default=0,
        ge=0,
        description="Total number of regulation violations",
    )


class PIIScanResponse(IDMixin, PIIScanSummary):
    """Full PII scan response with all details."""

    source_id: str = Field(..., description="Source that was scanned")
    status: Literal["pending", "running", "success", "failed", "error"] = Field(
        ...,
        description="Current scan status",
    )

    # Data statistics
    row_count: int | None = Field(default=None, description="Number of rows scanned")
    column_count: int | None = Field(default=None, description="Number of columns")

    # Scan configuration used
    min_confidence: float = Field(
        default=0.8,
        description="Confidence threshold used for this scan",
    )
    regulations_checked: list[str] | None = Field(
        default=None,
        description="Regulations that were checked",
    )

    # Findings (full details)
    findings: list[PIIFinding] = Field(
        default_factory=list,
        description="List of PII findings",
    )

    # Regulation violations
    violations: list[RegulationViolation] = Field(
        default_factory=list,
        description="List of regulation violations",
    )

    # Error info (if status is 'error')
    error_message: str | None = Field(
        default=None,
        description="Error message if scan failed",
    )

    # Timing
    duration_ms: int | None = Field(
        default=None,
        ge=0,
        description="Scan duration in milliseconds",
    )
    started_at: datetime | None = Field(default=None, description="Start timestamp")
    completed_at: datetime | None = Field(
        default=None,
        description="Completion timestamp",
    )
    created_at: datetime = Field(..., description="Record creation timestamp")

    @classmethod
    def from_model(cls, scan: Any) -> PIIScanResponse:
        """Create response from model.

        Args:
            scan: PIIScan model instance.

        Returns:
            PIIScanResponse instance.
        """
        findings = []
        if scan.result_json and "findings" in scan.result_json:
            findings = [
                PIIFinding(**finding) for finding in scan.result_json["findings"]
            ]

        violations = []
        if scan.result_json and "violations" in scan.result_json:
            violations = [
                RegulationViolation(**violation)
                for violation in scan.result_json["violations"]
            ]

        return cls(
            id=scan.id,
            source_id=scan.source_id,
            status=scan.status,
            total_columns_scanned=scan.total_columns_scanned or 0,
            columns_with_pii=scan.columns_with_pii or 0,
            total_findings=scan.total_findings or 0,
            has_violations=scan.has_violations or False,
            total_violations=scan.total_violations or 0,
            row_count=scan.row_count,
            column_count=scan.column_count,
            min_confidence=scan.min_confidence or 0.8,
            regulations_checked=scan.regulations_checked,
            findings=findings,
            violations=violations,
            error_message=scan.error_message,
            duration_ms=scan.duration_ms,
            started_at=scan.started_at,
            completed_at=scan.completed_at,
            created_at=scan.created_at,
        )


class PIIScanListItem(IDMixin, PIIScanSummary):
    """PII scan list item (without full findings/violations)."""

    source_id: str
    status: Literal["pending", "running", "success", "failed", "error"]
    row_count: int | None = None
    column_count: int | None = None
    min_confidence: float = 0.8
    regulations_checked: list[str] | None = None
    duration_ms: int | None = None
    created_at: datetime

    @classmethod
    def from_model(cls, scan: Any) -> PIIScanListItem:
        """Create list item from model.

        Args:
            scan: PIIScan model instance.

        Returns:
            PIIScanListItem instance.
        """
        return cls(
            id=scan.id,
            source_id=scan.source_id,
            status=scan.status,
            total_columns_scanned=scan.total_columns_scanned or 0,
            columns_with_pii=scan.columns_with_pii or 0,
            total_findings=scan.total_findings or 0,
            has_violations=scan.has_violations or False,
            total_violations=scan.total_violations or 0,
            row_count=scan.row_count,
            column_count=scan.column_count,
            min_confidence=scan.min_confidence or 0.8,
            regulations_checked=scan.regulations_checked,
            duration_ms=scan.duration_ms,
            created_at=scan.created_at,
        )


class PIIScanListResponse(ListResponseWrapper[PIIScanListItem]):
    """Paginated PII scan list response."""

    pass

File: frontend/src/mocks/factories/scan.ts
/**
 * PII scan factory - generates realistic PII scan results
 * Supports th.scan() parameters: columns, regulations, min_confidence
 */

import type { PIIScan, PIIFinding, RegulationViolation } from '@/api/client'
import {
  createId,
  createTimestamp,
  randomChoice,
  randomInt,
  faker,
} from './base'

/**
 * Options for th.scan() call - mirrors PIIScanRequest from API
 */
export interface PIIScanOptions {
  columns?: string[]
  regulations?: ('gdpr' | 'ccpa' | 'lgpd')[]
  min_confidence?: number
}

export interface PIIScanFactoryOptions {
  id?: string
  sourceId?: string
  status?: PIIScan['status']
  findingCount?: number
  hasViolations?: boolean
  violationCount?: number
  durationMs?: number
  rowCount?: number
  columnCount?: number
  minConfidence?: number
  regulationsChecked?: string[]
  /**
   * th.scan() options that influence mock behavior:
   * - columns: limits findings to these columns only
   * - regulations: determines violation generation
   * - min_confidence: filters findings below this threshold
   */
  options?: PIIScanOptions
}

// PII types commonly detected
const PII_TYPES = [
  'email',
  'phone',
  'ssn',
  'credit_card',
  'ip_address',
  'date_of_birth',
  'address',
  'name',
  'passport',
  'driver_license',
  'national_id',
  'bank_account',
]

// Columns that typically contain PII
const PII_COLUMNS = [
  'email',
  'user_email',
  'phone',
  'phone_number',
  'ssn',
  'social_security',
  'credit_card',
  'card_number',
  'ip_address',
  'ip',
  'birth_date',
  'dob',
  'address',
  'street_address',
  'full_name',
  'name',
  'first_name',
  'last_name',
  'passport_number',
  'driver_license',
  'national_id',
  'bank_account',
  'account_number',
]

// Regulation descriptions for violations
const REGULATION_MESSAGES: Record<string, string[]> = {
  gdpr: [
    'Personal data must be processed lawfully, fairly, and transparently',
    'Data must be collected for specified, explicit, and legitimate purposes',
    'Data must be adequate, relevant, and limited to what is necessary',
    'Data must be accurate and kept up to date',
    'Data must be kept in a form which permits identification for no longer than necessary',
    'Data must be processed securely',
  ],
  ccpa: [
    'Consumer has the right to know what personal information is collected',
    'Consumer has the right to delete personal information',
    'Consumer has the right to opt-out of sale of personal information',
    'Consumer has the right to non-discrimination for exercising their rights',
    'Business must provide notice at collection',
  ],
  lgpd: [
    'Processing must have a legal basis',
    'Data subject must be informed about data processing',
    'Data must be processed for legitimate purposes',
    'Data must be adequate and relevant for processing purposes',
    'Data security measures must be implemented',
  ],
}

const SEVERITIES: ('low' | 'medium' | 'high' | 'critical')[] = [
  'critical',
  'high',
  'medium',
  'low',
]

function createPIIFinding(options?: {
  column?: string
  piiType?: string
  minConfidence?: number
}): PIIFinding {
  const column = options?.column ?? randomChoice(PII_COLUMNS)
  const piiType = options?.piiType ?? randomChoice(PII_TYPES)
  const minConf = options?.minConfidence ?? 0.8
  // Generate confidence above threshold
  const confidence = Number(
    (minConf + Math.random() * (1 - minConf)).toFixed(2)
  )

  // Generate sample values based on PII type
  const sampleValues = generateSampleValues(piiType, randomInt(1, 5))

  return {
    column,
    pii_type: piiType,
    confidence,
    sample_count: randomInt(10, 10000),
    sample_values: sampleValues,
  }
}

function generateSampleValues(piiType: string, count: number): string[] {
  const generators: Record<string, () => string> = {
    email: () => faker.internet.email(),
    phone: () => faker.phone.number(),
    ssn: () => `***-**-${randomInt(1000, 9999)}`,
    credit_card: () => `****-****-****-${randomInt(1000, 9999)}`,
    ip_address: () => faker.internet.ip(),
    date_of_birth: () => faker.date.birthdate().toISOString().split('T')[0],
    address: () => faker.location.streetAddress(),
    name: () => faker.person.fullName(),
    passport: () => `***${faker.string.alphanumeric(5)}***`,
    driver_license: () => `***${faker.string.alphanumeric(6)}***`,
    national_id: () => `***${randomInt(1000000, 9999999)}***`,
    bank_account: () => `****${randomInt(1000, 9999)}`,
  }

  const generator = generators[piiType] || (() => faker.string.alphanumeric(10))
  return Array.from({ length: count }, generator)
}

function createRegulationViolation(options?: {
  regulation?: 'gdpr' | 'ccpa' | 'lgpd'
  column?: string
  piiType?: string
}): RegulationViolation {
  const regulation = options?.regulation ?? randomChoice(['gdpr', 'ccpa', 'lgpd'] as const)
  const column = options?.column ?? randomChoice(PII_COLUMNS)
  const piiType = options?.piiType ?? randomChoice(PII_TYPES)
  const messages = REGULATION_MESSAGES[regulation] || REGULATION_MESSAGES.gdpr
  const message = randomChoice(messages)

  return {
    regulation,
    column,
    pii_type: piiType,
    message,
    severity: randomChoice(SEVERITIES),
  }
}

function createFindings(
  count: number,
  options?: {
    columns?: string[]
    minConfidence?: number
  }
): PIIFinding[] {
  const findings: PIIFinding[] = []
  const availableColumns = options?.columns?.length
    ? options.columns
    : PII_COLUMNS

  for (let i = 0; i < count; i++) {
    findings.push(
      createPIIFinding({
        column: randomChoice(availableColumns),
        minConfidence: options?.minConfidence,
      })
    )
  }

  return findings
}

function createViolations(
  count: number,
  options?: {
    regulations?: ('gdpr' | 'ccpa' | 'lgpd')[]
    findings?: PIIFinding[]
  }
): RegulationViolation[] {
  const violations: RegulationViolation[] = []
  const regulations = options?.regulations?.length
    ? options.regulations
    : (['gdpr', 'ccpa', 'lgpd'] as const)

  for (let i = 0; i < count; i++) {
    const finding = options?.findings?.[i % (options?.findings?.length || 1)]
    violations.push(
      createRegulationViolation({
        regulation: randomChoice([...regulations]),
        column: finding?.column,
        piiType: finding?.pii_type,
      })
    )
  }

  return violations
}

export function createPIIScan(
  factoryOptions: PIIScanFactoryOptions = {}
): PIIScan {
  const { options: scanOptions, ...baseOptions } = factoryOptions
  const status = baseOptions.status ?? randomChoice(['success', 'failed', 'error'])

  // Determine finding count
  let findingCount: number
  if (baseOptions.findingCount !== undefined) {
    findingCount = baseOptions.findingCount
  } else {
    findingCount = status === 'error' ? 0 : randomInt(0, 15)
  }

  // Create findings
  const findings = createFindings(findingCount, {
    columns: scanOptions?.columns,
    minConfidence: scanOptions?.min_confidence ?? baseOptions.minConfidence ?? 0.8,
  })

  // Determine if has violations
  const hasViolations =
    baseOptions.hasViolations ??
    (scanOptions?.regulations && scanOptions.regulations.length > 0 && findingCount > 0)

  // Create violations if regulations were checked
  let violations: RegulationViolation[] = []
  if (hasViolations) {
    const violationCount =
      baseOptions.violationCount ?? Math.min(findingCount, randomInt(1, 5))
    violations = createViolations(violationCount, {
      regulations: scanOptions?.regulations,
      findings,
    })
  }

  // Calculate unique columns with PII
  const columnsWithPii = new Set(findings.map((f) => f.column)).size
  const totalColumnsScanned =
    baseOptions.columnCount ?? randomInt(columnsWithPii + 2, 50)

  const durationMs = baseOptions.durationMs ?? randomInt(500, 30000)
  const startedAt = createTimestamp(randomInt(0, 30))
  const completedAt =
    status !== 'error'
      ? new Date(new Date(startedAt).getTime() + durationMs).toISOString()
      : undefined

  return {
    id: baseOptions.id ?? createId(),
    source_id: baseOptions.sourceId ?? createId(),
    status,
    total_columns_scanned: totalColumnsScanned,
    columns_with_pii: columnsWithPii,
    total_findings: findings.length,
    has_violations: violations.length > 0,
    total_violations: violations.length,
    row_count: baseOptions.rowCount ?? randomInt(100, 1000000),
    column_count: totalColumnsScanned,
    min_confidence: scanOptions?.min_confidence ?? baseOptions.minConfidence ?? 0.8,
    regulations_checked:
      scanOptions?.regulations ?? baseOptions.regulationsChecked ?? null,
    findings,
    violations,
    error_message: status === 'error' ? faker.lorem.sentence() : undefined,
    duration_ms: durationMs,
    started_at: startedAt,
    completed_at: completedAt,
    created_at: startedAt,
  }
}

export function createPIIScans(count: number, sourceId?: string): PIIScan[] {
  return Array.from({ length: count }, () => createPIIScan({ sourceId })).sort(
    (a, b) =>
      new Date(b.created_at).getTime() - new Date(a.created_at).getTime()
  )
}

/**
 * Create diverse PII scans for comprehensive test coverage
 */
export function createDiversePIIScans(sourceId: string): PIIScan[] {
  const scans: PIIScan[] = []

  // 1. Clean scan - no PII found
  scans.push(
    createPIIScan({
      sourceId,
      status: 'success',
      findingCount: 0,
      hasViolations: false,
    })
  )

  // 2. PII found but no violations (no regulations checked)
  scans.push(
    createPIIScan({
      sourceId,
      status: 'success',
      findingCount: 5,
      hasViolations: false,
    })
  )

  // 3. GDPR violations
  scans.push(
    createPIIScan({
      sourceId,
      status: 'failed',
      findingCount: 8,
      hasViolations: true,
      options: {
        regulations: ['gdpr'],
      },
    })
  )

  // 4. Multiple regulation violations
  scans.push(
    createPIIScan({
      sourceId,
      status: 'failed',
      findingCount: 12,
      hasViolations: true,
      violationCount: 6,
      options: {
        regulations: ['gdpr', 'ccpa', 'lgpd'],
      },
    })
  )

  // 5. High confidence threshold scan
  scans.push(
    createPIIScan({
      sourceId,
      status: 'success',
      findingCount: 3,
      options: {
        min_confidence: 0.95,
      },
    })
  )

  // 6. Specific columns scan
  scans.push(
    createPIIScan({
      sourceId,
      status: 'success',
      findingCount: 4,
      options: {
        columns: ['email', 'phone', 'ssn'],
      },
    })
  )

  // 7. Error scan
  scans.push(
    createPIIScan({
      sourceId,
      status: 'error',
      findingCount: 0,
    })
  )

  // 8. Large dataset scan
  scans.push(
    createPIIScan({
      sourceId,
      rowCount: 10000000,
      columnCount: 150,
      durationMs: 120000,
    })
  )

  // 9. Additional random scans
  for (let i = 0; i < 5; i++) {
    scans.push(createPIIScan({ sourceId }))
  }

  return scans.sort(
    (a, b) =>
      new Date(b.created_at).getTime() - new Date(a.created_at).getTime()
  )
}
